include/pmacc/particles/algorithm/ForEach.hpp

```diff
@@ -22,157 +22,447 @@
+             * @return Lockstep foreach executor to operate on frames.
+             */
+            DINLINE auto lockstepForEach() const
+            {
+                return forEachParticleInFrame;
+            }
+
+            /** Construct algoritm to operate on particles in a superCell
              *
-             * @{
+             * @param workerIdx workerIdx index of the worker: range [0;workerSize)
+             * @param particlesBox particles memory
+             *                     It is not allowed concurrently to add or remove particles during the
+             *                     execution of this algorithm.
+             * @param superCellIdx index of the superCell where particles should be processed
              */
+            DINLINE ForEachParticle(
+                uint32_t const workerIdx,
+                T_ParBox const& particlesBox,
+                DataSpace<dim> const& superCellIdx)
+                : m_superCellIdx(superCellIdx)
+                , forEachParticleInFrame(workerIdx)
+                , m_particlesBox(particlesBox)
+            {
+            }
 
-            /** Version for a custom area mapper factory
+            /** Execute unary functor for each particle.
              *
-             * @tparam T_AreaMapperFactory factory type to construct an area mapper that defines the area to process,
-             *                             adheres to the AreaMapperFactory concept
+             * @attention There is no guarantee in which order particles will be executed.
+             *            If the particle functor is used the algorithm assumes that the frame structure is following
+             *            the rule tat all except the last frame are fully filled and in the last frame the low indices
+             *            within the frames will be contigious filled too.
              *
-             * @param areaMapperFactory factory to construct an area mapper,
-             *                          the area is defined by the constructed mapper object
+             * @tparam T_Acc alpaka accelerator type
+             * @tparam T_Functor unary particle functor or frame functor following the interface concept
+             *                           detail::FrameFunctorInterface or detail::ParticleFunctorInterface
+             * @param acc alpaka accelerator
+             * @param unaryFunctor Functor executed for each particle/frame.
+             *                     The caller must ensure that calling the functor in parallel with
+             *                     different workers is data race free.
+             *                     particle functor: 'void operator()(T_Acc const &, ParticleType)'.
+             *                       It is not allowed to call a synchronization function within the functor.
+             *                     frame functor: 'void operator()(T_Acc const &, FrameCtx)'.
```

Comment:
Did you mean `For frame ...` ?

Reply:
I will put a dash in front to separate that we handle once a particle functor and once the frame functor.